### PR TITLE
Chapter3

### PR DESCRIPTION
**Pull request by ramkumarr**
### Solutions for Chapter 3 (Without the final task :-/

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter3.hs [hunk 1]
```diff
@@ -374,6 +378,33 @@ after the fight. The battle has the following possible outcomes:
...
+    go :: Monster -> Knight -> String -> Either Int Knight
+    go Monster{mprops = Props{health = h, gold = g}} k@Knight{kprops = p} _
+      | h <= 0 = Right k{kprops=p{gold=gold p + g}}
+    go _ Knight{kprops = Props{health = h}} _ | h <= 0 = Left (-1)
+    go m@Monster{mprops = mp@Props{health = h}}
+       k@Knight{kprops = Props{attack = a}} "knight" =
+       go m{mprops = mp{health = h-a}} k "monster"
+    go m@Monster{mprops = Props{attack = a}}
+       k@Knight{kprops = kp@Props{health = h}} "monster" =
+       go m k{kprops = kp{health = h-a}} "knight"
```
Comment (chshersh) [inline src/Chapter3.hs:406]: Nice recursive solution!

File: src/Chapter3.hs [hunk 2]
```diff
@@ -480,6 +512,31 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data House = House Int deriving (Show) -- house with number of people in it
+data Church = Church deriving (Show)
+data Library = Library deriving (Show)
+data Castle = Castle String deriving (Show)
+data City = City {
+  castle :: Maybe Castle,
+  wall :: Bool,
```
Comment (chshersh) [inline src/Chapter3.hs:521]: This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

File: src/Chapter3.hs [hunk 3]
```diff
@@ -480,6 +512,31 @@ After defining the city, implement the following functions:
...
+buildHouse c@City{houses=h} num = c{houses=House num : h}
+
+buildWalls :: City -> Either String City
+buildWalls c =
+  case castle c of
+    Nothing -> Left "No castle in the city!"
+    Just _ ->
+      case foldl (\acc (House n) -> acc+n) 0 (houses c) of
+        n | n < 10 -> Left "Not enough people in the city!"
+        _ -> Right c{wall=True}
```
Comment (chshersh) [inline src/Chapter3.hs:539]: That's very good modelling of the result type with sum types. Great work :+1: 

If you want to improve the code even further, you can introduce a more precise custom data types:

```haskell
data BuildWallsResult
    = NoCastle
    | NotEnoughPeople
    | NewCity City
```

With a custom data type, it's more convenient to perform unit testing of various corner cases, rather than with comparing strings.

File: src/Chapter3.hs [hunk 4]
```diff
@@ -560,22 +617,30 @@ introducing extra newtypes.
...
 
-calculatePlayerDefense :: Int -> Int -> Int
-calculatePlayerDefense armor dexterity = armor * dexterity
+calculatePlayerDefense :: Armor -> Dexterity -> Defense
+calculatePlayerDefense (Armor a) (Dexterity d) = Defense (a*d)
 
-calculatePlayerHit :: Int -> Int -> Int -> Int
-calculatePlayerHit damage defense health = health + defense - damage
+calculatePlayerHit :: Damage -> Defense -> Health -> Health
+calculatePlayerHit (Damage dmg) (Defense def) (Health h) = Health (h+def-dmg)
```
Comment (chshersh) [inline src/Chapter3.hs:643]: Nice work on newtypes! :medal_sports: Newtypes are one very important feature in Haskell :+1:
You can read more about the newtype pattern in our blog post: https://kowainik.github.io/posts/haskell-mini-patterns#newtype

File: src/Chapter3.hs [hunk 5]
```diff
@@ -910,6 +988,21 @@ Implement instances of "Append" for the following types:
...
     append :: a -> a -> a
 
+newtype Gold = Gold Int
+instance Append Gold where
+    append :: Gold -> Gold -> Gold
+    append (Gold g1) (Gold g2) = Gold (g1+g2)
+
+instance Append [a] where
+    append :: [a] -> [a] -> [a]
+    append lst1 lst2 = lst1 ++ lst2
```
Comment (chshersh) [inline src/Chapter3.hs:998]: You can eta-reduce the implementation of `append` for brevity. So you can see how easy it is to create instances :slightly_smiling_face: 
```suggestion
    append = (++)
```

File: src/Chapter3.hs [hunk 6]
```diff
@@ -910,6 +988,21 @@ Implement instances of "Append" for the following types:
...
+
+instance (Append a) => Append (Maybe a) where
+    append :: Maybe a -> Maybe a -> Maybe a
+    append Nothing Nothing = Nothing
+    append Nothing (Just x) = Just x
+    append (Just x) Nothing = Just x
```
Comment (chshersh) [inline src/Chapter3.hs:1004]: This is an absolutely correct solution!

However, we can notice that these 3 cases could be reduced to 2!
Let's see what we can unite in here.

If some of the arguments are `Nothing`, you actually always return the second one! So you need to inspect only two cases: `Nothing` in the first position, and `Nothing` on the second one 🙂

File: src/Chapter3.hs [hunk 7]
```diff
@@ -971,6 +1064,38 @@ implement the following functions:
...
+nextDay Saturday = Sunday
+
+daysToParty :: WeekDay -> Int
+daysToParty Sunday = 5
+daysToParty Monday = 4
+daysToParty Tuesday = 3
+daysToParty Wednesday = 2
+daysToParty Thursday = 1
+daysToParty Friday = 0
+daysToParty Saturday = 6
```
Comment (chshersh) [inline src/Chapter3.hs:1097]: These functions don't use standard functions of standard typeclasses. I would recommend looking into `Bounded` and `Enum` typeclasses, derive them and try to see what of their methods you can use :slightly_smiling_face:

**Review by chshersh — APPROVED** [review]
Nice work on Chapter Three so far! :+1: 

I left my review with some comments and suggestions :slightly_smiling_face:

**Comment by chshersh** [comment]
Thanks for participating in Learn4Haskell! Hope you enjoyed this Hacktoberfest 🥳 

I am going to close this PR as we can't merge the solutions into our course 🔐 

Even though the Hacktoberfest is over, we are going to continue reviewing, so feel free to open new PRs, if you haven't finished the course yet! 🎊
